onnxruntime/core/providers/cpu/tensor/gather_elements.cc

```diff
@@ -135,23 +136,37 @@ static void core_impl(const Tensor* input_tensor, const Tensor* indices_tensor,
   auto num_elements = indices_tensor->Shape().Size();
   int64_t lower_index_limit = -input_shape[axis];
   int64_t upper_index_limit = input_shape[axis] - 1;
-  for (int64_t i = 0; i < num_elements; ++i) {
+
+  auto validation_fn = [indices_data, lower_index_limit, upper_index_limit](ptrdiff_t i) {
     auto indices_val = indices_data[i];
     if (indices_val < lower_index_limit || indices_val > upper_index_limit)
       ORT_THROW("GatherElements op: Value in indices must be within bounds [",
                 lower_index_limit, " , ", upper_index_limit, "]. Actual value is ", indices_val);
+  };
+  for (int64_t i = 0; i < num_elements; ++i) {  // TODO: parallelize this? didn't give any benefit in my tests
+    validation_fn(i);
   }
 
   int64_t num_inner_dim = calculate_num_inner_dim(indices_shape);
   int64_t inner_dim_size = indices_shape[input_rank - 1];
   bool processing_inner_dim = (axis == input_rank - 1) ? true : false;
 
   int64_t base_offset = 0;
-  Tin indices_counter = -1;
-  int64_t output_counter = -1;
+  Tin indices_counter = 0;
   size_t element_size = input_tensor->DataType()->Size();
 
   std::vector<int64_t> process_dims(input_rank, 0);
+  int64_t output_counter = 0;
+
+  auto conditional_batch_call = [ttp, inner_dim_size](std::function<void(ptrdiff_t)> f) {
+    if (inner_dim_size < 10 * 1000) {  // TODO: tune this, arbitrary threshold
```

Comment:
Do we need a simple test for both the ops that invokes the parallel conditional branch (i.e.) >= 10k elements ?

Reply:
sure, will do.